File: lib/native_is_for_everyone/distro_managers/dnf.dart
import 'package:process_run/process_run.dart';
import 'package:xpm/native_is_for_everyone/models/native_package.dart';
import 'package:xpm/native_is_for_everyone/native_package_manager.dart';
import 'package:xpm/os/executable.dart';

/// Package manager adapter for Fedora/RHEL systems using DNF
class DnfPackageManager extends NativePackageManager {
  @override
  Future<List<NativePackage>> search(String name, {int? limit}) async {
    final shell = Shell(verbose: false);
    final result = await shell.run('dnf search "$name" 2>/dev/null');

    if (result.first.exitCode != 0) {
      return [];
    }

    final packages = _parseSearchOutput(result.first.stdout.toString());

    if (limit != null) {
      return packages.take(limit).toList();
    }

    return packages;
  }

  List<NativePackage> _parseSearchOutput(String output) {
    final packages = <NativePackage>[];
    final lines = output.split('\n');

    // Skip header lines (first few lines are usually headers)
    bool headerPassed = false;

    for (final line in lines) {
      if (line.isEmpty) continue;

      // Skip header lines
      if (line.contains('===') || line.contains('Matched') || line.startsWith('Last metadata')) {
        headerPassed = true;
        continue;
      }

      if (!headerPassed) continue;

      // DNF output format: "package-name.arch : Description"
      final colonIndex = line.indexOf(' : ');
      if (colonIndex == -1) continue;

      final nameArch = line.substring(0, colonIndex).trim();
      final description = line.substring(colonIndex + 3).trim();

      // Extract package name (remove .arch suffix)
      final dotIndex = nameArch.lastIndexOf('.');
      final name = dotIndex > 0 ? nameArch.substring(0, dotIndex) : nameArch;
      final arch = dotIndex > 0 ? nameArch.substring(dotIndex + 1) : null;

      packages.add(NativePackage(
        name: name,
        description: description,
        arch: arch,
        repo: 'fedora',
      ));
    }

    return packages;
  }

  @override
  Future<void> install(String name) async {
    final sudo = await Executable('sudo').find();
    if (sudo == null) {
      throw Exception('sudo not found');
    }
    final shell = Shell();
    await shell.run('$sudo dnf install -y "$name"');
  }

  @override
  Future<bool> isInstalled(String name) async {
    final shell = Shell(verbose: false);
    final result = await shell.run('rpm -q "$name" 2>/dev/null');
    return result.first.exitCode == 0;
  }

  @override
  Future<NativePackage?> get(String name) async {
    final shell = Shell(verbose: false);
    final result = await shell.run('dnf info "$name" 2>/dev/null');

    if (result.first.exitCode != 0) {
      return null;
    }

    return _parsePackageInfo(name, result.first.stdout.toString());
  }

  NativePackage _parsePackageInfo(String packageName, String output) {
    String? version;
    String? description;
    String? arch;
    String? repo;

    for (final line in output.split('\n')) {
      final colonIndex = line.indexOf(':');
      if (colonIndex == -1) continue;

      final field = line.substring(0, colonIndex).trim().toLowerCase();
      final value = line.substring(colonIndex + 1).trim();

      switch (field) {
        case 'version':
        case 'versão':
          version = value;
          break;
        case 'summary':
        case 'description':
        case 'descrição':
          description = value;
          break;
        case 'architecture':
        case 'arquitetura':
        case 'arch':
          arch = value;
          break;
        case 'from repo':
        case 'repository':
        case 'repositório':
          repo = value;
          break;
      }
    }

    return NativePackage(
      name: packageName,
      version: version,
      description: description,
      arch: arch,
      repo: repo,
    );
  }
}

File: crates/xpm-core/src/native_pm/dnf.rs
//! DNF package manager integration (Fedora/RHEL)

use super::{NativePackage, NativePackageManager};
use crate::os::executable::Executable;
use anyhow::{Context, Result};
use async_trait::async_trait;
use std::path::PathBuf;
use tokio::process::Command;

pub struct DnfPackageManager {
    dnf_path: PathBuf,
    sudo_path: Option<PathBuf>,
}

impl DnfPackageManager {
    pub async fn new() -> Self {
        let dnf_path = Executable::new("dnf")
            .find()
            .unwrap_or_else(|| PathBuf::from("/usr/bin/dnf"));

        let sudo_path = Executable::new("sudo").find();

        Self { dnf_path, sudo_path }
    }

    async fn run_dnf(&self, args: &[&str]) -> Result<String> {
        let output = Command::new(&self.dnf_path)
            .args(args)
            .output()
            .await
            .context("Failed to run dnf command")?;

        if output.status.success() {
            Ok(String::from_utf8_lossy(&output.stdout).to_string())
        } else {
            let stderr = String::from_utf8_lossy(&output.stderr);
            anyhow::bail!("dnf command failed: {}", stderr)
        }
    }

    async fn run_dnf_sudo(&self, args: &[&str]) -> Result<String> {
        let mut cmd = if let Some(sudo) = &self.sudo_path {
            let mut cmd = Command::new(sudo);
            cmd.arg(&self.dnf_path);
            cmd
        } else {
            Command::new(&self.dnf_path)
        };

        let output = cmd
            .args(args)
            .output()
            .await
            .context("Failed to run dnf command")?;

        if output.status.success() {
            Ok(String::from_utf8_lossy(&output.stdout).to_string())
        } else {
            let stderr = String::from_utf8_lossy(&output.stderr);
            anyhow::bail!("dnf command failed: {}", stderr)
        }
    }

    fn parse_search_output(&self, output: &str) -> Vec<NativePackage> {
        let mut packages = Vec::new();

        for line in output.lines() {
            if line.trim().is_empty() || line.starts_with('=') || line.starts_with("Last metadata") {
                continue;
            }

            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() >= 2 {
                let name_arch = parts[0];
                let name = name_arch.split('.').next().unwrap_or(name_arch);
                let version = parts.get(1).map(|s| s.to_string());

                packages.push(NativePackage {
                    name: name.to_string(),
                    version,
                    description: None,
                    arch: name_arch.split('.').nth(1).map(|s| s.to_string()),
                    repo: parts.get(2).map(|s| s.to_string()),
                    popularity: None,
                    installed: false,
                });
            }
        }

        packages
    }
}

#[async_trait]
impl NativePackageManager for DnfPackageManager {
    fn name(&self) -> &str {
        "dnf"
    }

    async fn search(&self, query: &str, limit: Option<usize>) -> Result<Vec<NativePackage>> {
        let output = self.run_dnf(&["search", query]).await?;
        let mut packages = self.parse_search_output(&output);

        if let Some(limit) = limit {
            packages.truncate(limit);
        }

        Ok(packages)
    }

    async fn install(&self, name: &str) -> Result<()> {
        self.run_dnf_sudo(&["install", "-y", name]).await?;
        Ok(())
    }

    async fn remove(&self, name: &str) -> Result<()> {
        self.run_dnf_sudo(&["remove", "-y", name]).await?;
        Ok(())
    }

    async fn is_installed(&self, name: &str) -> Result<bool> {
        let output = Command::new("rpm")
            .args(["-q", name])
            .output()
            .await?;

        Ok(output.status.success())
    }

    async fn get(&self, name: &str) -> Result<Option<NativePackage>> {
        let output = self.run_dnf(&["info", name]).await;

        match output {
            Ok(output) => {
                let mut pkg = NativePackage::new(name);

                for line in output.lines() {
                    if let Some((key, value)) = line.split_once(':') {
                        let value = value.trim();
                        match key.trim() {
                            "Version" => pkg.version = Some(value.to_string()),
                            "Summary" | "Description" => pkg.description = Some(value.to_string()),
                            "Architecture" => pkg.arch = Some(value.to_string()),
                            _ => {}
                        }
                    }
                }

                pkg.installed = self.is_installed(name).await.unwrap_or(false);
                Ok(Some(pkg))
            }
            Err(_) => Ok(None),
        }
    }

    async fn update_db(&self) -> Result<()> {
        self.run_dnf_sudo(&["makecache"]).await?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_search_output() {
        let pm = DnfPackageManager {
            dnf_path: PathBuf::from("/usr/bin/dnf"),
            sudo_path: None,
        };

        let output = "vim-enhanced.x86_64    2:9.0.1000-1.fc38    fedora\n\
                      vim-minimal.x86_64    2:9.0.1000-1.fc38    fedora";

        let packages = pm.parse_search_output(output);

        assert_eq!(packages.len(), 2);
        assert_eq!(packages[0].name, "vim-enhanced");
        assert_eq!(packages[0].version, Some("2:9.0.1000-1.fc38".to_string()));
        assert_eq!(packages[0].arch, Some("x86_64".to_string()));
    }

    #[test]
    fn test_parse_search_output_skips_headers() {
        let pm = DnfPackageManager {
            dnf_path: PathBuf::from("/usr/bin/dnf"),
            sudo_path: None,
        };

        let output = "Last metadata expiration check: 0:30:00 ago\n\
                      ======================== Name Matched ========================\n\
                      git.x86_64    2.40.0-1.fc38    updates";

        let packages = pm.parse_search_output(output);

        assert_eq!(packages.len(), 1);
        assert_eq!(packages[0].name, "git");
    }

    #[test]
    fn test_name() {
        let pm = DnfPackageManager {
            dnf_path: PathBuf::from("/usr/bin/dnf"),
            sudo_path: None,
        };
        assert_eq!(pm.name(), "dnf");
    }
}
